eclipse/sw360
# feat(ClearingRequestUI): Bug fixes and enhancement

[//]: # (Copyright Bosch.IO GmbH 2020)
[//]: # (This program and the accompanying materials are made)
[//]: # (available under the terms of the Eclipse Public License 2.0)
[//]: # (which is available at https://www.eclipse.org/legal/epl-2.0/)
[//]: # (SPDX-License-Identifier: EPL-2.0)

> Please provide a summary of your changes here.
> Which issue is this pull request belonging to and how is it solving it? (*Refer to issue here*)
* #1247 & #1248 

> Did you add or update any new dependencies that are required for your change?
* No

Issue: #1247 & #1248 

### Suggest Reviewer
> You can suggest reviewers here with an @mention.

### How To Test?
> How should these changes be tested by the reviewer?
1. Testing for issue #1247
    * `Clearing request` can now be closed, even if `Agreed clearing date` is passed.
    * Create a `Clearing request`, set the `Agreed clearing date` to some previous date in database (for reproducing the issue)
Now you'll be able to `Close` the `CR`.

2. Testing for issue #1248
    * Added a new field `Priority` (`enum`) in clearing request, with following `enum` values:
        1. LOW
        2. MEDIUM
        3. HIGH
        4. CRITICAL
    * Removed (hidden) the `clearing team` field from `open clearing request` table and added the new field `Priority` (as shown in screenshot 1).
    * `Priority` field is displayed as dropdown in `edit mode`, and it can be set/modified only by `clearing team`. 
    * Enabled `Requesting User comment` option while re-opening the `closed` / `rejected` CR.
        1. This new comment will be appended with the `CR` reopen comment. (as shown in screenshot 2).
    * Changed the logic of selecting the `Preferred clearing date` while creating / re-opening the `CR`, such that.
        1. The number of future days to be disabled can be configured by `Admin` in `Liferay Expando API` custom field section.
        2. The new field name is `PreferredClearingDateLimit`.
        3. If field is not configured then it is added programmatically with empty value.
        4. If the field is value is not configured by `Admin`, then default value of 7 days is considered. In such case, `requesting user` can select `preferred clearing date` greater than 7 days from creation of `CR.`
        5. `PreferredClearingDateLimit` field value can be configured from `Global Menu` icon on top right --> `Control Panel` --> `Custom Fields` --> `User` --> `PreferredClearingDateLimit`. and then change it's value.
    * Added a `Critical` checkbox with appropriate info  message in `Create CR` and `Reopen CR` dialog box (as shown in screenshot 3)
        1. The `Critical` checkbox will be shown if there are less than 2 critical `CR`.
        2. The `Critical` checkbox will *NOT* be shown, if there are more than 2 critical `CR`.
    * Changes is `Clearing Team` field in Edit mode will be logged as a comment.
    * Test that `priority` filter for `clearing request` table is working fine.

### Screenshot 1:
![image](https://user-images.githubusercontent.com/50870174/121105726-44952380-c822-11eb-86b5-ec623cc9bf31.png)

### Screenshot 2:
![image](https://user-images.githubusercontent.com/50870174/121130729-6efbd680-c84c-11eb-9234-ed5a35be8b01.png)


### Screenshot 3:
![image](https://user-images.githubusercontent.com/50870174/121107250-31378780-c825-11eb-86a6-bb1de6f9435a.png)


### Checklist
Must:

Signed-off-by: Abdul Kapti <abdul.kapti@siemens-healthineers.com>
